app/models/service_template_transformation_plan_task.rb

```diff
@@ -98,6 +99,10 @@ def preflight_check_vm_exists_in_destination_openstack
     end
   end
 
+  def two_phase?
+    source.snapshots.empty?
```

Comment:
@fdupont-redhat so a two-phase migration will always be done if there are no snapshots on the source vm?

Reply:
@agrare, not exactly. A warm migration will also be a two-phase migration, but it's limited to warm migration plans. But, the restriction is the the same from a VM perspective: it must not have snapshots.

When a user creates a migration plan, the UI asks the infrastructure mapping for a list of VM that match the mapping. In the result, there's a field to tell if the VM is eligible to warm migration: https://github.com/ManageIQ/manageiq/blob/master/app/models/transformation_mapping/vm_migration_validator.rb#L165

This support feature mixin is currently only implemented in VMware provider and it returns true if the VM has no snapshot: https://github.com/ManageIQ/manageiq-providers-vmware/blob/master/app/models/manageiq/providers/vmware/infra_manager/vm.rb#L13-L15

The difference with warm migration is that there is a fall back mechanism, one-phase migration, which is the historical implementation.

We verify snapshot presence, because they block CBT activation if it is not yet enabled on the VM. And two-phase cold migration currently uses a CBT snapshot, because the code reuses most of warm migration. This will likely change in the future to remove that limitation.

Reply:
So is this method indicating that this transformation_plan_task _is_ two-phase or _could be_ two-phase?

Reply:
Two-phase performs better than one-phase, so we want to make it the default whenever possible. So, it means that the transformation_plan_task _is_ two-phase.

Reply:
Okay, "so a two-phase migration will always be done if there are no snapshots on the source vm?" :smile: